File: src/EditorFeatures/CSharpTest/Structure/NamespaceDeclarationStructureTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CSharp.Structure;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Structure;
using Microsoft.CodeAnalysis.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Structure;

[Trait(Traits.Feature, Traits.Features.Outlining)]
public sealed class NamespaceDeclarationStructureTests : AbstractCSharpSyntaxNodeStructureTests<NamespaceDeclarationSyntax>
{
    internal override AbstractSyntaxStructureProvider CreateProvider() => new NamespaceDeclarationStructureProvider();

    [Fact]
    public Task TestNamespace()
        => VerifyBlockSpansAsync("""
                class C
                {
                    {|hint:$$namespace N{|textspan:
                    {
                    }|}|}
                }
                """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestNamespaceWithLeadingComments()
        => VerifyBlockSpansAsync("""
                class C
                {
                    {|span1:// Goo
                    // Bar|}
                    {|hint2:$$namespace N{|textspan2:
                    {
                    }|}|}
                }
                """,
            Region("span1", "// Goo ...", autoCollapse: true),
            Region("textspan2", "hint2", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestNamespaceWithNestedUsings()
        => VerifyBlockSpansAsync("""
                class C
                {
                    {|hint1:$$namespace N{|textspan1:
                    {
                        {|hint2:using {|textspan2:System;
                        using System.Linq;|}|}
                    }|}|}
                }
                """,
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false),
            Region("textspan2", "hint2", CSharpStructureHelpers.Ellipsis, autoCollapse: true));

    [Fact]
    public Task TestNamespaceWithNestedUsingsWithLeadingComments()
        => VerifyBlockSpansAsync("""
                class C
                {
                    {|hint1:$$namespace N{|textspan1:
                    {
                        {|span2:// Goo
                        // Bar|}
                        {|hint3:using {|textspan3:System;
                        using System.Linq;|}|}
                    }|}|}
                }
                """,
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false),
            Region("span2", "// Goo ...", autoCollapse: true),
            Region("textspan3", "hint3", CSharpStructureHelpers.Ellipsis, autoCollapse: true));

    [Fact]
    public Task TestNamespaceWithNestedComments()
        => VerifyBlockSpansAsync("""
                class C
                {
                    {|hint1:$$namespace N{|textspan1:
                    {
                        {|span2:// Goo
                        // Bar|}
                    }|}|}
                }
                """,
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false),
            Region("span2", "// Goo ...", autoCollapse: true));
}

File: src/EditorFeatures/VisualBasicTest/Structure/NamespaceDeclarationStructureTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports Microsoft.CodeAnalysis.Structure
Imports Microsoft.CodeAnalysis.VisualBasic.Structure
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Outlining
    <Trait(Traits.Feature, Traits.Features.Outlining)>
    Public Class NamespaceDeclarationStructureProviderTests
        Inherits AbstractVisualBasicSyntaxNodeStructureProviderTests(Of NamespaceStatementSyntax)

        Friend Overrides Function CreateProvider() As AbstractSyntaxStructureProvider
            Return New NamespaceDeclarationStructureProvider()
        End Function

        <Fact>
        Public Async Function TestNamespace() As Task
            Const code = "
{|span:$$Namespace N1
End Namespace|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Namespace N1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestNamespaceWithComments() As Task
            Const code = "
{|span1:'My
'Namespace|}
{|span2:$$Namespace N1
End Namespace|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "' My ...", autoCollapse:=True),
                Region("span2", "Namespace N1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestNamespaceWithNestedComments() As Task
            Const code = "
{|span1:$$Namespace N1
{|span2:'My
'Namespace|}
End Namespace|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "Namespace N1 ...", autoCollapse:=False),
                Region("span2", "' My ...", autoCollapse:=True))
        End Function
    End Class
End Namespace
